rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+- Whoever is running tests (user or automation) will still require IAM permissions in order to test modules
+
+#### Dev Deployment
+
+- Dev deployment requires prerequisite deployment of IAM resource module
+- Engineers do not require IAM permissions to deploy changes to dev
+
+#### Prod Deployment
+
+- Prod deployment requires prerequisite deployment of IAM resource module
+- Engineers do not require IAM permissions to deploy changes to prod
+
+#### Risks
+
+- Onerous process to deploy IAM modules separately from application modules
+- Module testing requires creating IAM resources separately
+- Engineers can still define IAM resources anywhere, though they'll lack the permissions to apply them
+
+# Prior art
+[prior-art]: #prior-art
+
+<!--
+Discuss prior art, both the good and the bad, in relation to this proposal. A few examples of what
+this can include are:
+
+- Can we learn something about this proposal from other projects in CTD?
+- Can we learn something about this proposal from other transit agencies?
+- Can we learn something about this proposal from past experiences in other jobs or projects?
+
+This section is intended to encourage you as an author to think about the lessons from other places,
+and provide readers of your RFC with a fuller picture.
+
+If there is no prior art, that is fine.
+
+Note that while precedent is some motivation, it does not on its own motivate an RFC.
+-->
+
+Workflow automation for Terraform is entirely new territory for the authors of this document. We've done a fair amount of research and factfinding to formulate a plan that we think brings CTD's usage of Terraform more in line with industry best practices, and takes advantage of products that take some of the work off our shoulders. That said, it's entirely possible there's something we missed, and we're open to recommendations of solutions, software or patterns from anyone who cares to weigh in.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+<!--
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature
+  before stabilization?
+- What related issues do you consider out of scope for this RFC that could be addressed in the
+  future independently of the solution that comes out of this RFC?
+-->
+
+One major unresolved question is the timeline for implementing all the improvements laid out in this RFC. As noted in the [Drawbacks][drawbacks] section above, it's a lot of work. Once this RFC is approved we'll need to scope out the work and divide it accordingly between the infrastructure and engineering teams.
+
+Other questions for the reader:
+
+- Is the proposed [root module naming convention][root-module-naming-convention] sensible and sufficiently comprehensive?
+- Does the [isolated module testing workflow][reference-module-testing] make sense? Does it seem easy enough to carry out?
```

Comment:
One thing I think I missed in it is how do you know if it worked? If it completed without errors? Are there health checks or assertions to make? Would it be usable for application end-to-end tests?

Reply:
This might be context-dependent, but I think "completed without errors" will be a good baseline for most modules. Assertions are an interesting idea, though I'm not sure how we'd carry that out exclusively in Terraform. ([this third-party provider might be an option.](https://registry.terraform.io/providers/bwoznicki/assert/latest/docs))

I do think end-to-end tests are something that _could_ be possible, but I don't want to go too deep in recommending strategies that are specific to a certain type of application.

I'll see if I can add more to the workflow doc to clarify this.

Reply:
https://github.com/mbta/technology-docs/pull/16/commits/80af5b572a6dbd4a614fa314f5b37756dd348a5b